### PR TITLE
Use `Gc::drop` to prevent early finalization

### PR DESCRIPTION
**Pull request by jacob-hughes**
This commit adds a memory barrier to `Gc`'s drop which will prevent the compiler from optimising away the reference too soon. This is a special implementation with compiler support, because it is usually impossible to allow both `Drop` and `Copy` traits to be implemented on a type simultaneously.

The remove_gc_drops pass is then used to remove drop calls in Gc where a barrier is not needed (such as for temporaries and non-finalizable types).

### CODE REVIEW

File: compiler/rustc_mir_transform/src/remove_gc_drops.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                if !ty.is_gc(tcx) {
+                    continue;
+                }
+
+                if let ty::Adt(_, substs) = ty.kind() {
+                    if is_gc_crate
+                        || !decl.is_user_variable()
+                        || !substs.type_at(0).needs_finalizer(tcx, param_env)
+                    {
+                        debug!("SUCCESS: replacing `drop` with goto({:?})", target);
```
Comment (ltratt) [inline compiler/rustc_mir_transform/src/remove_gc_drops.rs:36]: Should we remove this?

File: library/alloc/src/gc.rs [hunk 2]
```diff
@@ -98,6 +98,36 @@ impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<Gc<U>> for Gc<T> {}
...
+/// free up the register for something else. This is fine (and usually
+/// desirable!) because it doesn't have any observable difference in behaviour.
+///
+/// However, things get complicated when a garbage collector is involved. In
+/// very rare cases, this optimisation, followed by an unfortunately timed
+/// collection, may cause the value the reference points to to be freed earlier
+/// than expected - and thus finalized earlier than it should be. This can cause
+/// deadlocks, races, and even use-after-frees.
+///
+/// Adding a memory barrier to `Gc`'s drop will prevent the compiler from
```
Comment (ltratt) [inline library/alloc/src/gc.rs:116]: s/will prevent/prevents/

File: library/alloc/src/gc.rs [hunk 3]
```diff
@@ -98,6 +98,36 @@ impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<Gc<U>> for Gc<T> {}
 #[unstable(feature = "gc", issue = "none")]
 impl<T: ?Sized + Unsize<U>, U: ?Sized> DispatchFromDyn<Gc<U>> for Gc<T> {}
 
+/// A memory barrier to prevent finalizers running before the last reference to
```
Comment (ltratt) [inline library/alloc/src/gc.rs:101]: "Barriers" are both a compile-time and a run-time concept. This is a "compiler" barrier, right?

**Comment by ltratt** [comment]
Could we clarify what "too soon" means? I think it means "until the end of the block scope"?

**Comment by ltratt** [comment]
Oops, didn't mean to press the "close" button!

File: library/alloc/src/gc.rs (hunk 3 above)
Comment (jacob-hughes) [reply library/alloc/src/gc.rs:101]: That's right. Do you think it'd be clearer if I changed this from memory barrier to compiler barrier?

File: library/alloc/src/gc.rs (hunk 2 above)
Comment (jacob-hughes) [reply library/alloc/src/gc.rs]: Fixed: 337e4e498ec8cdf8472ee140ffdbf8f46ce30e57

File: compiler/rustc_mir_transform/src/remove_gc_drops.rs (hunk 1 above)
Comment (jacob-hughes) [reply compiler/rustc_mir_transform/src/remove_gc_drops.rs]: Sure, fixed: 929778e9bec5f3a5b689affd15945b997f63f1c8

File: library/alloc/src/gc.rs (hunk 3 above)
Comment (ltratt) [reply library/alloc/src/gc.rs:101]: I do indeed :)

**Comment by ltratt** [comment]
> Yeah "too soon" here means before the end of the block scope. I can reword this commit as part of a squash?

Works for me! Let's do the "compiler barrier" fix first though.

**Comment by jacob-hughes** [comment]
Fixed: 6d4a51eccff181994163775bc1124737b0fafa5b

**Comment by jacob-hughes** [comment]
Squashed

**Comment by jacob-hughes** [comment]
The linkage for liballoc is a bit weird, and things that apply to lang_items (such as `gc`) sometimes have to be marked as `not(test)` to prevent them from being exported twice. I've done this for `Drop` now, which has fixed the build issue, and it still passes the new early finalizer test. (6c62eb31a5aaeb29065a2fd64d2b3694e7395873)

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**